wait: Introduce new methods that allow detection of context cancellation

smarterclayton opened the pull request:
The Poll* methods predate context in Go, and the current implementation will return ErrWaitTimeout even if the context is cancelled, which prevents callers who are using Poll* from handling that error directly(for instance, if you want to cancel a function in a controlled fashion but still report cleanup errors to logs, you want to know the difference between 'didn't cancel', 'cancelled cleanly', and 'hit an error).  This blocks us from adding context cancellation in the Kubelet so that pods react more quickly to interruption and so grace periods can be shortened after the fact on long-running grace period pods.

This PR adds two new methods that reflect how modern Go uses context in polling while preserving all Kubernetes-specific behavior:
    
        PollUntilContextCancel
        PollUntilContextTimeout
    
These methods can be used for infinite polling (normal context), timed polling (deadline context), and cancellable poll (cancel context). All other Poll/Wait methods are marked as deprecated for removal in 1.28 (or a chosen future release). The ErrWaitTimeout error will no longer be returned from the `Poll*` methods, but will continue to be returned from `ExponentialBackoff*`. Users updating to use these new methods are responsible for converting their error handling as appropriate. A convenience helper `Interrupted(err) bool` has been added that should be used instead of checking `err == ErrWaitTimeout`. In a future release ErrWaitTimeout will be made private to prevent incorrect use. The helper can be used with all current polling methods since context cancellation and deadline are semantically equivalent to ErrWaitTimeout. A new `ErrorInterrupted(cause error)` method should be used instead of returning ErrWaitTimeout in custom code.

The convenience method PollUntilContextTimeout is added because deadline context creation is verbose and the cancel function must be called to properly cleanup the context - many of the current poll users would see code sizes increase at the call site. To reduce the overall method surface area, the distinction between PollImmediate and Poll has been reduced to a single boolean on PollUntilContextCancel so we do not need multiple helper methods.
    
The existing methods were not altered because ecosystem callers have been observed to use ErrWaitTimeout to mean "any error that my condition func did not return" which prevents cancellation errors from being returned from the existing methods.

Callers migrating to `PollUntilContextCancel` should:

1. Pass a context with a deadline or timeout if they were previously using `Poll*Until*` and check `err` for `context.DeadlineExceeded` instead of `ErrWaitTimeout` (more specific) or use `Interrupted(err)` for a generic check.
2. Callers that were waiting forever or for context cancellation should ensure they are checking `context.Canceled` instead of `ErrWaitTimeout` to detect when the poll was stopped early.

Callers of `ExponentialBackoffWithContext` should use `Interrupted(err)` instead of directly checking `err == ErrWaitTimeout`.  No other changes are needed.

Code that returns `ErrWaitTimeout` should instead define a local cause and return `wait.ErrorInterrupted(cause)`, which will be recognized by `wait.Interrupted()`. If nil is passed the previous message will be used but clients are highly recommended to use typed checks vs message checks.

The `NewExponentialBackoffManager` and `NewJitteringBackoffManager` methods have been marked deprecated and callers may either use the new `Timer()` method on the BackoffManager to pass to `BackoffUntil` until they are removed, or create a `Backoff` struct and invoke `DelayWithReset(clock, interval).Timer(clock)` now.  Callers that pass a zero reset interval are advised to invoke Timer() on an appropriate `Backoff` struct instead, since it is more efficient.

These new methods are a prerequisite for supporting context cancellation of Kubelet `Sync*Pod` methods, because wait.X eat cancellation errors and make it impossible to determine whether the error was early exit or failure.  It will also reduce the surface area and make our use of context idiomatic for new users and reduce accidental checks.

Callers within Kubernetes will be bulk updated in a follow on PR that will include this commit so people can see the changes in use.

**NOTE**: This pull was part of a series of changes that introduce new context-cancellation-aware Poll methods, reduce the surface area of the wait package to a smaller set of functions (if unused, marked private, if consolidating, marked deprecated), unify the underlying loop implementation with better testing, consolidate the backoff manager code into a smaller chunk, and in general address a number of outstanding issues.  See #116476, #115077, #115116, #115113, #115140, #115064, and #107826.

A future refactor will update the old implementations to use the new, more efficient `loopConditionUntilContext` once it has been tested more.

/kind cleanup

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
The `wait.Poll*` and `wait.ExponentialBackoff*` functions have been deprecated and will be removed in a future release.  Callers should switch to using `wait.PollUntilContextCancel`, `wait.PollUntilContextTimeout`, or `wait.ExponentialBackoffWithContext` as appropriate.

`PollWithContext(Cancel|Deadline)` will no longer return `ErrWaitTimeout` - use the `Interrupted(error) bool` helper to replace checks for `err == ErrWaitTimeout`, or compare specifically to context errors as needed. A future release will make the `ErrWaitTimeout` error private and callers must use `Interrupted()` instead. If you are returning `ErrWaitTimeout` from your own methods, switch to creating a location specific `cause err` and pass it to the new method `wait.ErrorInterrupted(cause) error` which will ensure `Interrupted()` returns true for your loop. 

The `wait.NewExponentialBackoffManager` and `wait.NewJitteringBackoffManager` functions have been marked as deprecated.  Callers should switch to using the `Backoff{...}.DelayWithReset(clock, resetInterval)` method and must set the `Steps` field when using `Factor`. As a short term change, callers may use the `Timer()` method on the `BackoffManager` until the backoff managers are deprecated and removed. Please see the godoc of the deprecated functions for examples of how to replace usage of this function.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

smarterclayton commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107826/pull-kubernetes-e2e-kind-ipv6/1486804124145029120

```
Kubernetes  e2e suite: [sig-auth] ServiceAccounts should set ownership and  permission when RunAsUser or FsGroup is present [LinuxOnly]  [NodeFeature:FSGroup] expand_less | 6m33s
-- | --
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/auth/service_accounts.go:488 Jan 27 21:21:50.183: Unexpected error:     <*errors.errorString \| 0xc003256170>: {         s: "expected pod \"test-pod-c9ec94e9-eb29-49a6-b493-d5512bb22756\" success: Gave up after waiting 5m0s for pod \"test-pod-c9ec94e9-eb29-49a6-b493-d5512bb22756\" to be \"Succeeded or Failed\"",     }     expected pod "test-pod-c9ec94e9-eb29-49a6-b493-d5512bb22756" success: Gave up after waiting 5m0s for pod "test-pod-c9ec94e9-eb29-49a6-b493-d5512bb22756" to be "Succeeded or Failed" occurred /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/framework/util.go:770
```

probably not this, but

/retest

smarterclayton commented:
Blocks https://github.com/kubernetes/kubernetes/pull/107829 in some cases - any wait.Poll (like volume mounts) wraps the error.

fedebongio commented:
/assign @lavalamp 
Daniel, we thought you may have some thoughts on changing this API. 
/triage accepted

smarterclayton commented:
I’ll note the godoc on the WithContext methods imply a behavior they do not actually fully implement, which mostly defeats the point of context cancelation:

> context.Canceled will be returned if the ctx.Done() channel is closed without fn ever returning true.

The Context methods inconsistently return these methods - this commit makes WithContext behave the way we document, and leaves the existing behavior unchanged.

I struggle to imagine someone using WaitWithContext who wants ErrWaitTimeout when the context is canceled (since it’s not a timeout).  However, if I missed something here we can always declare bankruptcy and go to WaitWithContext2

lavalamp commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go:
```diff
@@ -939,7 +939,7 @@ func TestPollImmediateUntilWithContext(t *testing.T) {
 				defer cancel()
 				return ctx, cancel
 			},
-			errExpected:      ErrWaitTimeout,
```
Did you search through existing uses to see if any need to be changed?

smarterclayton replied:
No broken uses in our tree.  However, there are a few broken users on github.  I think for that reason I'll create new methods and mark the old ones deprecated.  Opinion on naming scheme to create the new correct ones?

Scenario that would be impacted:

1. User calls PollImmediateUntilWithContext() with a condition that returns no error
2. User checks the error value and *requires* it to equal ErrWaitTimeout (vs handling err != nil as a catch all), otherwise it does "something different"
3. The behavior of that "something different" is such that any other error besides ErrWaitTimeout means the condition failed with an error
4. The condition failing with the error vs being cancelled has a different meaning.

If the user simply handles the error generically this has no impact.  If the user treats any error except ErrWaitTimeout as "stop early", this has no impact.  Only if the user assumes `err != nil && err != ErrWaitTimeout`  means "the condition errored" will this have an impact.  That should be rare, but it's definitely possible.

Search https://cs.github.com/?scopeName=All+repos&scope=&q=%22.PollImmediateWithContext%22+ErrWaitTimeout had 11 hits, 6 of which are the same code:

* https://cs.github.com/alebedev87/node-observability-operator/blob/6c749f38b63cd2a9b63c158c27314d7d63812a9a/pkg/operator/controller/machineconfig/mcp.go?q=%22.PollImmediateWithContext%22+ErrWaitTimeout#L257 does not check for non ErrWaitTimeout errors, so a context cancellation would be treated as a nil error.

smarterclayton replied:
Considering `*WithContextCancellable`.

smarterclayton commented:
This came up with @dashpole's change to context propagation again, reopening this and doing another pass on it to make sure we get this correct (so we can propagate context up and out of wait methods correctly).

dims commented:
@smarterclayton do we want to do this in 1.26 itself? (code freeze deadline tuesday i think)

dims commented:
/assign @dashpole 
/remove-lifecycle rotten
/priority important-soon

/milestone v1.27

(since we only fix tests or bugs at this point)

smarterclayton commented:
Will target with a rename in 1.27 and attempt to land

aojea commented:
> Will target with a rename in 1.27 and attempt to land

what about wrapping the error and using errors.Is() or As() in the kubelet?

smarterclayton commented:
> what about wrapping the error and using errors.Is() or As() in the kubelet?

Now that contexts are widely used in Go, it's safer for us to align our usage with the behavior users would expect.  There are other places where context cancellation will be useful, such as controllers and safe shutdown.  Rather than put a quick fix in, I'd rather align our overall codebase to something long term useful.  Also, the key bug here is that Poll doesn't *always* return context.Canceled, which means sometimes sync will log spurious errors, and that will put extra burden on admins.

smarterclayton commented:
@tkashem as per slack - please review if you can.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -203,8 +203,75 @@ func Jitter(duration time.Duration, maxFactor float64) time.Duration {
+//	if wait.EndedEarly(err) {
+//	    log.Infof("Wait for operation exceeded")
+//	} else ...
+func EndedEarly(err error) bool {
```
❤️

aojea replied:
name doesn't ring well though

smarterclayton replied:
I was playing off the `wait.EndedEarly(err)` - still doesn't work for you?  Suggestions?

aojea replied:
I mean, this error is one of these ones: 
- errErrWaitTimeout
- context.Canceled
- context.DeadlineExceeded

all represent that something was interrupted either because it took too long or it was canceled.

Meanwhile I write this I got it, is "ended earlier than expected", but it may end early for a good thing ... what about `ErrAborted` and `Aborted` or ErrWaitAborted , OpAborted, that emphasize the fact that something was going on and was cancelled

smarterclayton replied:
Note I'll likely rename errErrWaitTimeout to match whatever this ends up being.  Let me think some more on your suggestions and look at some examples in more recent error handling in the stdlib to see if I can spot some other suggestions.

smarterclayton replied:
How about `wait.Interrupted(err error) bool` - if the wait was interrupted by something other than the condition (cancel, deadline, or possibly panic)?

aojea replied:
that sgtm

smarterclayton commented:
#115064 is also attempting to clean up another part of wait - not a prereq to this, but will be a prereq to cleaning up `Until`.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -418,19 +480,128 @@ func (j *jitteredBackoffManagerImpl) Backoff() clock.Timer {
+			after.Reset(waitBeforeRetry)
+		}
+
+		select {
+		case <-after.C:
+		case <-ctx.Done():
+			return ctx.Err()
+		}
```
it feels like this `select block` and the one at the beginning of the loop could be merged somehow ... but I can't say since I don't quite understand well the lazy initialization of the  timer, just thinking out loud

smarterclayton replied:
Well spotted - there are some tricky behaviors in the various loop methods in this package, and this is one of them.  Specifically Backoff currently attempts to guarantee that once a context is cancelled you get at most one more invocation of the function, which because of the randomness of go select casing this construct doesn't guarantee, so you have to check twice.  I'm going to tweak the code slightly so this looks more like BackoffUntil - as a consequence of that we'll give that guarantee to all our methods.  During shutdown that sort of guarantee is useful to avoid excessive error logging that might confuse a user.

smarterclayton replied:
See the more coherent (and now used across all three types of waits in this package) `loopConditionUntilContext` - has some comments inline.

ash2k commented:
Perhaps https://github.com/kubernetes/utils is a better place for polling helpers? I doesn't have anything related to api or api types, it's probably in apimachinery for historical reasons? In this repo it even sits under `util` - `k8s.io/apimachinery/pkg/util/wait`.

p.s. happy to see this refactoring!

smarterclayton commented:
/retest

smarterclayton commented:
We've been a bit hesitant to move stuff to utils - it should really be things that are "done" and are unlikely to change in the future - the refactoring here indicates they aren't quite there yet.  Also, with 800 references I selfishly prefer to keep them colocated in k/k while we do the rollout (to simplify issues) - perhaps afterwards they may be candidates.

smarterclayton commented:
/retest

aojea commented:
Also `k8s.io/utils` is an external repo and had been problematic for bugfixes, to fix something in k/k you have to PR to k8s.io/utils without the possibility of testing on k/k until you revendor again (of course you can do all this dance manually, but that puts much load on developers for small win)

CecileRobertMichon commented:
/test pull-kubernetes-e2e-capz-conformance

trying to debug test above as last few failures are inconsistent ^

smarterclayton commented:
The unit failure is because CPU load, trying to figure out how to effectively test it.  Doesn't block anything (test issue, not change issue)

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -317,103 +337,46 @@ func (b *Backoff) Step() time.Duration {
-// NewExponentialBackoffManager returns a manager for managing exponential backoff. Each backoff is jittered and
-// backoff will not exceed the given max. If the backoff is not called within resetDuration, the backoff is reset.
-// This backoff manager is used to reduce load during upstream unhealthiness.
-func NewExponentialBackoffManager(initBackoff, maxBackoff, resetDuration time.Duration, backoffFactor, jitter float64, c clock.Clock) BackoffManager {
```
this change affects a lot of projects https://grep.app/search?q=NewExponentialBackoffManager, can't we keep both constructor and change only the internals?

smarterclayton replied:
Let me review (using https://grep.app/search?q=wait.NewExponentialBackoffManager&case=true&words=true)

The signature change is roughly:

```
- bm := wait.NewExponentialBackoffManager(time.Millisecond, maxBackoff, 100*time.Millisecond, 2.0, 1.0, fakeClock)
+ bm := wait.Backoff{Duration: time.Millisecond, Cap: maxBackoff, Factor: 2.0, Jitter: 1.0, Steps: math.MaxInt}.DelayWithReset(fakeClock, 100*time.Millisecond)
```

and then at point of use you call `.Timer(<your clock|wait.RealTimer>)` OR you invoke the delay func.  I am slightly worried about the mechanical change (people have to get the right arguments to Backoff) but the impact to point of use is low.

That being said there is no reason we have to remove the constructor inside the codebase - we can mark it deprecated and take feedback.  I'll make that change, thanks.

smarterclayton replied:
See my other question about whether to have a helper on the existing manager and delay when they have to get rid of the manager.  I suspect the answer should be "unless we remove the manager constructor now, just add the helper to bridge the gap".

smarterclayton replied:
(which is what i implemented)

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -288,17 +307,18 @@ type Backoff struct {
 		}
 		return b.Duration
+	} else {
+		b.Steps--
```
this is not really needed, right? the previous if always return
I think this is more confusing , since next instructions seems to follow a different flow than b.Step

smarterclayton replied:
The else?  Yeah, should be removed now.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -132,24 +131,39 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
 //
 // If sliding is true, the period is computed after f runs. If it is false then
 // period includes the runtime for f.
-func BackoffUntil(f func(), backoff BackoffManager, sliding bool, stopCh <-chan struct{}) {
```
this change looks less disruptive https://grep.app/search?q=wait.BackoffUntil&filter[lang][0]=Go

smarterclayton replied:
The open question with me deprecating instead of removing `New*` is whether we need to add a transition method on the existing manager.  If we do, users of wait.BackoffUntil have to change twice in the future (once now to use that helper and then once later when the constructors are removed).  If we don't add the helper, they have to change now but only do so once.  Both are disruptive.  Changing twice makes this easier to land but more work for them overall.

Preference?

smarterclayton replied:
I added the helper and pushed that change.  I'll note that callers can choose to use that until the manager is deprecated.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -132,24 +131,39 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
+	RealTimer = clock.RealClock{}.NewTimer
+)
+
+// TimerFunc allows the caller to inject a timer for testing. Most callers should use RealTimer.
```
if that is the case should we even expose it?
or just make it private to override only on these tests?

aojea replied:
Timer also says https://pkg.go.dev/time#Timer

>  A Timer must be created with NewTimer or AfterFunc.

that may open another unsafe options for users , or can they create this `clock.Timer` without `NewTimer` ?

smarterclayton replied:
Good point, I think TimerFunc can be safely removed now (delay func and Timer interface cover it).

smarterclayton replied:
Updated in latest.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -158,7 +172,12 @@ func BackoffUntil(f func(), backoff BackoffManager, sliding bool, stopCh <-chan
+		if t == nil {
+			t = timerFn(interval)
+		} else {
+			t.Reset(interval)
```
https://pkg.go.dev/time#Timer.Reset

> For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels.

does this meet this requirement?

smarterclayton replied:
To get to this point in the function, we had to have waited on T's channel at least once.  All other paths exit the function.

Note: you *don't* have to drain a channel for it to get garbage collected, you only have to Stop it (that's a VERY subtle thing that was tripping up people before).

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -132,24 +131,39 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
+		if t != nil {
+			t.Stop()
+		}
+	}()
```
wow, this was a subtle bug, if the previous code returned on the first select it will never stop the timer

smarterclayton replied:
It *should* lead to a goroutine leak, but i'm not 100% positive.  We didn't seem to see signs of it but we also don't use a ton of wait loops that exit except at cleanup.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/timer.go:
```diff
@@ -0,0 +1,110 @@
+		return
+	}
+	d := t.fn()
+	t.t.Reset(d)
```
similar comment as in https://github.com/kubernetes/kubernetes/pull/107826#discussion_r1084295000

> For a Timer created with NewTimer, Reset should be invoked only on stopped or expired timers with drained channels.

does this means that, `Next()` can only be called if the timer has stopped or the channel was drained?

or that the Next() should wait?
```
if !t.t.Stop() {
	<-t.t.C
}
t.t.Reset(d)
```

smarterclayton replied:
Updated the Godoc of Timer, let me know if this clearer.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -288,33 +298,74 @@ type Backoff struct {
+// delayFunc returns a function that will compute the next interval to
+// wait given the arguments in b. It does not mutate the original backoff
+// but the function is safe to use only from a single goroutine.
+func (b Backoff) delayFunc() DelayFunc {
```
we should choose value or pointer receiver for Backoff, it seems value is ok since we don't modify the type,  but the `b.Step()` method in line294 is defined as receiver

aojea replied:
I defer to someone with more knowledge on the golang internals, I just flagged  it because I couldn't find if is correct or no,  only found this reference  https://go.dev/tour/methods/8

> In general, all methods on a given type should have either value or pointer receivers, but not a mixture of both. (We'll see why over the next few pages.)

smarterclayton replied:
Reason to have the separation

1. Existing method needs pointer receiver to work (backward compat for all existing code, no breakages)
2. New methods are non-mutating and deliberately designed to be easily inlined: `Backoff{Step: 3, Duration: time.Millisecond}.Timer()`.  If we required a pointer, all inline call sites get uglier `(&Backoff{Step: 3, Duration: time.Millisecond}).Timer()` which is more work for the user and more confusing.

Sufficiently justified?

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go:
```diff
@@ -379,6 +395,60 @@ func TestPollError(t *testing.T) {
+		time.Sleep(time.Microsecond)
+	}
+
+	fakeClock.Step(time.Second - time.Millisecond)
+	if attempt != 0 {
+		t.Fatalf("should still be waiting for condition")
+	}
+	fakeClock.Step(2 * time.Millisecond)
```
ok, so you check here the boundaries ,999ms and 1,001 s

aojea commented:
From the maximum of my capabilities this overall LGTM (just 2 comments)

However, I recognize that I lost track on some of the tests, and you should get another pair of eyes to do a sanity check

smarterclayton commented:
Thanks for the reviews Antonio, I'll probably wire some code up to the new methods to test for safety (the e2e have been fairly happy so far, but not for the new methods).

smarterclayton commented:
/test all

ash2k commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:
```diff
@@ -132,49 +132,33 @@ func NonSlidingUntilWithContext(ctx context.Context, f func(context.Context), pe
+// BackoffUntil loops until stop channel is closed, run f every duration given by delayFn.
+// An appropriately delayFn is provided by the Backoff.Step method. If sliding is true, the
+// period is computed after f runs. If it is false then period includes the runtime for f.
+func BackoffUntil(f func(), timer Timer, sliding bool, stopCh <-chan struct{}) {
```
May I suggest to move `f` to be the last argument? Code is more readable if `f` is a closure. And, maybe, `stopCh` should be the first argument, like a context? I assume it's not a context to be usable from the existing code?

smarterclayton replied:
This PR tries to minimize the changes to existing signatures except where unavoidable.  Instead of `wait.BackoffUtil` you can use `wait.Backoff{...}.DelayFunc().Until(ctx, bool, bool, func ...)`.  BackoffUntil should probably be removed in a future release, but I didn't mark it deprecated here.

smarterclayton commented:
/retest 

integration failure in https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107826/pull-kubernetes-integration/1628844207135264768 doesn't appear to be related because client-go doesn't call wait except in the test path, so a context cancellation error isn't because of the changes here.

aojea commented:
/test pull-kubernetes-integration

@smarterclayton only test changes since https://github.com/kubernetes/kubernetes/pull/107826#issuecomment-1407517114 ?

smarterclayton commented:
Correct, only adjusting a new test that came in.

aojea commented:
no flake

/test pull-kubernetes-integration


there was a revert recently because it flaked the integration testing, maybe this hit that but let's give it other round to be sure

smarterclayton commented:
/test all

smarterclayton commented:
Capz failures appear to be flaking in main and are not specific to this PR https://storage.googleapis.com/k8s-triage/index.html?text=Kubernetes%20cluster%20unreachable

CecileRobertMichon commented:
> Capz failures appear to be flaking in main and are not specific to this PR

That flake should have been already addressed in the release-1.7 branch... Here is the periodic job running the same test on k8s main branch, there are no recent flakes: https://testgrid.k8s.io/provider-azure-master-signal#capz-release-conformance

let's retry one more time and if it fails again I will dig into the failures

/retest

smarterclayton commented:
I'm reviewing disk capz failures - there's the possibility that the changes to Backoff return a different error, which disk attach failures are handling differently (instead of retry, they stop, perhaps?).  However, while I see a context deadline exceeded error on a few of those failures to attach, I don't see the azure controller using anything wait.ExponentialBackoff, which doesn't set a deadline.  Trying to find who is setting the deadline that causes the executor to fail.

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
@CecileRobertMichon, I see:

```
  There were additional failures detected after the initial failure:
    [PANICKED]
    Test Panicked
    In [AfterEach] at: /usr/local/go/src/runtime/panic.go:260
    runtime error: invalid memory address or nil pointer dereference
    Full Stack Trace
      sigs.k8s.io/azuredisk-csi-driver/test/e2e.glob..func2.2()
      	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:64 +0x73
```

in a passing run, is that known for capz?  https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107826/pull-kubernetes-e2e-capz-azure-disk-vmss/1630994819222343680

CecileRobertMichon commented:
This is part of the Azure CSI drivers e2e tests, @andyzhangx please take a look

andyzhangx commented:
> @CecileRobertMichon, I see:
> 
> ```
>   There were additional failures detected after the initial failure:
>     [PANICKED]
>     Test Panicked
>     In [AfterEach] at: /usr/local/go/src/runtime/panic.go:260
>     runtime error: invalid memory address or nil pointer dereference
>     Full Stack Trace
>       sigs.k8s.io/azuredisk-csi-driver/test/e2e.glob..func2.2()
>       	/home/prow/go/src/sigs.k8s.io/azuredisk-csi-driver/test/e2e/pre_provisioning_test.go:64 +0x73
> ```
> 
> in a passing run, is that known for capz? https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107826/pull-kubernetes-e2e-capz-azure-disk-vmss/1630994819222343680

looks like rerun the test `pull-kubernetes-e2e-capz-azure-disk-vmss`, it's pass now.

smarterclayton commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/107826/pull-kubernetes-e2e-capz-windows-containerd/1630994819117486080 is in master https://storage.googleapis.com/k8s-triage/index.html?date=2023-02-27&pr=1&text=Provisioning&job=capz#ae964a4730497553500e

/retest

smarterclayton commented:
/test all

smarterclayton commented:
flake was general (not specific to this change)

/retest

smarterclayton commented:
Integration failure appears to be specific to this PR

liggitt commented:
> Integration failure appears to be specific to this PR

nope, longstanding gap in our etcd3 driver not recognizing context cancelled errors when wrapped in an RPC error. exposed by https://github.com/kubernetes/kubernetes/pull/115096 which meant the test was talking directly to etcd, fixed by https://github.com/kubernetes/kubernetes/pull/116393 an hour ago.

smarterclayton commented:
/retest

None of the other tests appear specific to this PR

furkatgofurov7 commented:
Hi @smarterclayton!

K8s 1.27 Bug Triage Shadow here. I see a lot of activities happening on this PR, however FYI, with Code Freeze approaching next week on March 15th, if you would like this PR to be part of the v1.27 release, please make sure it goes in before above mentioned date.
Post Code Freeze, we will only be able to accept Release Blockers and Critical issues. Thank you!

liggitt reviewed (commented):
I made it about 20% of the way through the convert / unify commits, but had a _really_ hard time following the rewrite to verify the behavior was either identical or intentionally different and more correct

Left a couple ancillary comments, but if you want more feedback from me I'm afraid it'll take a solid day or two devoted to reviewing this :-/

smarterclayton commented:
Proposed after discussion on slack - pure additive (new error, new poll), mark old methods deprecated, separate PR for 1.28 that changes their impl (for longer soak)?

smarterclayton commented on staging/src/k8s.io/client-go/tools/watch/until_test.go:
```diff
@@ -209,7 +209,7 @@ func TestUntilWithSync(t *testing.T) {
 				return true, nil
 			},
-			expectedErr:   errors.New("timed out waiting for the condition"),
+			expectedErr:   wait.ErrWaitTimeout,
```
Is that unfortunate a "it's ok if it's minor and one time and the outcome is clearly better", or "we should just never change the type of ErrWaitTimeout"?

smarterclayton replied:
I feel like if a go package exposed `context.Canceled` and we were comparing it to `errors.New("context canceled")` and then go changed the impl type, it wouldn't fall within the compat guarantees of golang stdlib?  Although not certain.

smarterclayton committed:
wait: Deprecate legacy Poll methods for new context aware methods

The Poll* methods predate context in Go, and the current implementation
will return ErrWaitTimeout even if the context is cancelled, which
prevents callers who are using Poll* from handling that error directly
(for instance, if you want to cancel a function in a controlled fashion
but still report cleanup errors to logs, you want to know the difference
between 'didn't cancel', 'cancelled cleanly', and 'hit an error).

This commit adds two new methods that reflect how modern Go uses
context in polling while preserving all Kubernetes-specific behavior:

	PollUntilContextCancel
	PollUntilContextTimeout

These methods can be used for infinite polling (normal context),
timed polling (deadline context), and cancellable poll (cancel context).
All other Poll/Wait methods are marked as deprecated for removal in
the future. The ErrWaitTimeout error will no longer be returned from the
Poll* methods, but will continue to be returned from ExponentialBackoff*.
Users updating to use these new methods are responsible for converting
their error handling as appropriate. A convenience helper
`Interrupted(err) bool` has been added that should be used instead of
checking `err == ErrWaitTimeout`. In a future release ErrWaitTimeout will
be made private to prevent incorrect use. The helper can be used with all
polling methods since context cancellation and deadline are semantically
equivalent to ErrWaitTimeout. A new `ErrorInterrupted(cause error)` method
should be used instead of returning ErrWaitTimeout in custom code.

The convenience method PollUntilContextTimeout is added because deadline
context creation is verbose and the cancel function must be called to
properly cleanup the context - many of the current poll users would see
code sizes increase. To reduce the overall method surface area, the
distinction between PollImmediate and Poll has been reduced to a single
boolean on PollUntilContextCancel so we do not need multiple helper methods.

The existing methods were not altered because ecosystem callers have been
observed to use ErrWaitTimeout to mean "any error that my condition func
did not return" which prevents cancellation errors from being returned
from the existing methods. Callers must make a deliberate migration.

Callers migrating to `PollWithContextCancel` should:

1. Pass a context with a deadline or timeout if they were previously using
	`Poll*Until*` and check `err` for `context.DeadlineExceeded` instead of
	`ErrWaitTimeout` (more specific) or use `Interrupted(err)` for a generic
	check.
2. Callers that were waiting forever or for context cancellation should
	ensure they are checking `context.Canceled` instead of `ErrWaitTimeout`
	to detect when the poll was stopped early.

Callers of `ExponentialBackoffWithContext` should use `Interrupted(err)`
instead of directly checking `err == ErrWaitTimeout`. No other changes are
needed.

Code that returns `ErrWaitTimeout` should instead define a local cause
and return `wait.ErrorInterrupted(cause)`, which will be recognized by
`wait.Interrupted()`. If nil is passed the previous message will be used
but clients are highly recommended to use typed checks vs message checks.

As a consequence of this change the new methods are more efficient - Poll
uses one less goroutine.

smarterclayton committed:
wait: Update tests to reference the constant error

The structure of the error is changing, and we don't guarantee
reflect.DeepEqual(...) will remain true for ErrWaitTimeout currently.

smarterclayton commented:
Split #116476 out (moves files) then reduced this to just additive changes and net new tests (some minor test cleanups for context).  All existing code paths are identical.  The impl has been reviewed by @aojea.  In a future release the impl of the old methods could be removed or replaced (for ExponentialBackoff and the Managers).  

Is this more tractable to review?

smarterclayton commented:
Discussed with jordan, if pure additive (modulo the change in type of error) he's ok with someone else on final review - @aojea do you have time between now and code freeze?

smarterclayton commented:
/retest

Pull QPS failure in an e2e, unrelated

smarterclayton commented:
/retest

smarterclayton commented:
/retest

smarterclayton commented:
/retest

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/wait_test.go:
```diff
@@ -779,45 +801,137 @@ func TestExponentialBackoffManagerWithRealClock(t *testing.T) {
+func TestBackoffDelayWithResetEmpty(t *testing.T) {
+	fc := testingclock.NewFakeClock(time.Now())
+	backoff := Backoff{Duration: 1, Cap: 10, Factor: 2.0, Jitter: 0.0, Steps: 10}.DelayWithReset(fc, 0)
+	durations := []time.Duration{1, 1, 1, 1, 1, 1, 1}
```
it took me a while, this resets everytime is executed

smarterclayton replied:
I'll comment.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -0,0 +1,445 @@
+			},
+		},
+		{
+			name:      "condition success",
```
just comments for review, no need to address

"condition success and immediate"

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -0,0 +1,445 @@
+			},
+		},
+		{
+			name:    "condition success",
```
"condition success and sliding"

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -0,0 +1,445 @@
+	return b
+}
+
+func Test_loopConditionUntilContext_timings(t *testing.T) {
```
I'm not able to understand this test :(

It seems you've been using it during development to test something , because have some arguments that are not used like `context`or `errExpected`or always with the same value like  `callback`

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -0,0 +1,445 @@
+	return b
+}
+
+func Test_loopConditionUntilContext_timings(t *testing.T) {
```
I'm not able to understand this test :(

It seems you've been using it during development to test something , because have some arguments that are not used like `context`or `errExpected`or always with the same value like  `callback`

smarterclayton replied:
We can't run the timing test in a CPU constrained env, but I wanted to be certain the tests actually work, so when single threaded on my laptop i can see what output is.

smarterclayton replied:
I'll add a comment to the function.

smarterclayton replied:
Wrong test, heh.

This test records the delays calculated by the top level arguments and also advances the clock based on what the timer gets invoked with.  So basically validating both the outputs of DelayFunc and comparing them to what the loop passed to NewTimer and ensuring they matched in some basic scenarios.

It was to verify that DelayFunc got invoked.

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/loop_test.go:
```diff
@@ -0,0 +1,445 @@
+	// whether the OS or go runtime scheduler wake the timers, excess duration
+	// is logged by default and can be converted to a fatal error for testing.
+	// fail := t.Fatalf
+	fail := t.Logf
```
ok, this is noop in CI

aojea commented:
almost there

aojea commented on staging/src/k8s.io/apimachinery/pkg/util/wait/delay.go:
```diff
@@ -0,0 +1,51 @@
+// Timer takes an arbitrary delay function and returns a timer that can handle arbitrary interval changes.
+// Use Backoff{...}.Timer() for simple delays and more efficient timers.
+func (fn DelayFunc) Timer(c clock.Clock) Timer {
+	return &variableTimer{fn: fn, new: c.NewTimer}
```
this is the most complex thing to understand to me, so DelayFunc can generate a Timer that is a vriableTimer

However, variableTimer uses the `new`constructor to generate the Timer
```

type variableTimer struct {
	fn  DelayFunc
	t   clock.Timer
	new func(time.Duration) clock.Timer
}

func (t *variableTimer) C() <-chan time.Time {
	if t.t == nil {
		d := t.fn()
		t.t = t.new(d)
	}
	ret
```

smarterclayton replied:
Yeah, if you have an arbitrary delay func, you can't provide an optimized implementatnion (time.NewTicker for non-sliding is more efficient than continuously creating a timer.  A timer is "one use to get a channel" - so it's a lazy init.  When you Stop a timer you can then reuse it.  We're wrapping clock.Timer which wraps time.Timer.

So variableTimer is the inefficient version because DelayFunc hides arbitrary delays.

aojea commented:
/lgtm

I the existing behavior seems preserved to me.
Add. a new bunch of methods with some more specialized ways of playing with time.

@smarterclayton you need someone from cloud-provider to approves because of the unit tests modified there, maybe @andrewsykim can help you there

Thanks

smarterclayton commented:
/retest

dims commented:
/approve

(the legacy cloud provider stuff looks good, straight-forward)

/hold - please remove hold when you are ready

smarterclayton commented:
/hold cancel
